introduce a new synthetic hostname "_outbound" that maps to "the" local IP address

poettering opened the pull request:
This is supposed to be a shortcut for "the" local IP address, i.e. the most likely local address used for communication with other hosts. It's determined by asking the kernel for a routing decision towards the default gateway and then using whatever is used as local address for that decision.

This mostly scratches one of my own itches, that in order to figure out "the" local IP address on some host in a reasonably way you have to guess or deal with a wall of text from the `ip` tool.

bluca commented on src/shared/local-addresses.c:
```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+                        log_debug_errno(r, "Failed to bind socket to interface %i, ignoring: %m", a[i].ifindex);
+
+                if (connect(fd, &sa.sa, SOCKADDR_LEN(sa)) < 0)
+                        log_debug_errno(errno, "Failed to connect SOCK_DGRAM socket to gateway, ignoring: %m");
```
is there any point in continuing with `getsockname()` if `connect()` fails?

poettering replied:
My thinking here was that connect() might ultimately fail for various reasons, for example firewall/policy decisions, there could be anything. But if the routing decision was already made when that failed, then we can still use it. Hence, after calling connect() we check with getsockname() if it is bound now. if so, we use the data regardles if connect() failed. If it's still unbound then we detect that and return a failure.

i.e. i figured it would increase the chance for the whole logic to work, as we can neatly recognize late rwhether the routing decision was made or not.

i'll add a comment about that

bluca commented on src/shared/local-addresses.c:
```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+                done_af = a[i].family;
+        }
+
+        if (ret) {
```
is there any point in doing the work if `ret` is not set? iow, shouldn't it be asserted?

poettering replied:
local_addresses() + locale_gateways() which work very similar support being called with NULL arguments too. We implement basically the same API, hence we should do so here too. And it might even make sense to call the function with ret=NULL, for example if all you want to check is if we have a routable local IP address or not. if we do the function will return > 0 after all, if we don't we'll return == 0. And that info alone is very useful I'd say.

(But of course, we don#t actually call it this way)

bluca replied:
ok, makes sense to me

bluca commented on src/resolve/resolved-dns-synthesize.c:
```diff
@@ -321,17 +326,17 @@ static int synthesize_gateway_rr(Manager *m, const DnsResourceKey *key, int ifin
         af = dns_type_to_af(key->type);
         if (af >= 0) {
-                n = local_gateways(m->rtnl, ifindex, af, &addresses);
+                n = lookup(m->rtnl, ifindex, af, &addresses);
```
shouldn't there be an assert on `lookup`? (maybe not, getting late and might be confused)

poettering replied:
yeah, we should add that. done

bluca reviewed (approved):
Looks good to me, just a couple of minor questions. I have a feeling that `_outbound` will not be very stable with multiple ifaces, especially if a host has more complex routing than just static/dhcp derived, but then again it's a new standalone special hostname so it doesn't hurt one bit to have.

keszybz commented on src/shared/local-addresses.c:
```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+         * address off the socket that was chosen for the routing decision.
+         *
+         * This will generate exactly 0, 1 or 2 addresses depending if AF_INET, AF_INET6 is
+         * requested + configured. */
```
Hmm, why cut off the other addresses (see "here" below)? In the common scenario, there is just one address of a given family that is preferred, so this doesn't matter. But when we *are* in a setup where there is more than one address, the kernel will "randomly" select one of the routes and use it for a specific destination address. But for another connection to a different target IP it might select a different source address. If we select one of the addresses ourselves, we are not reporting the full information to the user… So this selection needs to be either justified, or dropped.

poettering replied:
My original usecase for this was to return "*the*" IP address of the local machine, i.e. the primary one that would actually be used for outbound traffic in the likeliest of cases. i.e. it's supposed to answer the question "what is my IP address" as well as we possibly can. 

Now if we'd return a list of multiple addresses here then this would defeat the point partially as the answer we provide would not be "*the*" IP address anymore, but a possible series of them.

Moreover there's a pretty high chance if we'd iterate through all defined gateways we'd always come to the same conclusion, i.e. the same IP address, over and over again, and would then just filter that out again and spend work on it needlessly.

I mean, this is not useful in truly multi-homed systems with multiple default gateways and many local IP addresses anyway, I have no illusions about this. This is supposed to be useful only for those cases where things are simple, i.e. there is one default gateway, and one IP address that is used to talk to it. And these simple cases are probably 99% of the installations I guess, which is good enough for me.

Or to say this differently: if you have multiple default gateways and multiple local IP addresses that are equally suited to talk to them, then `_outbound` is not your thing anyway, and there's value in leaving our search loop early, since this stuff might not be for free if there myriads of souch gateways or IP addresses defined.

Does that make sense?

I'll add a comment about this to the sources.

keszybz replied:
You make it sound like we could do the right thing in more cases.

> Moreover there's a pretty high chance if we'd iterate through all defined gateways we'd always come to the same conclusion, i.e. the same IP address, over and over again, and would then just filter that out again and spend work on it needlessly.

Well, sure, but how much that "extra work" will be? Except very rare cases, people have just a small handful of routes, so the extra connect() is completely negligible.

And then, if we get the same answer in all cases, that's great. We return a single address and the user is happy.
OTOH, if there *are* multiple addresses, and we return multiple addresses, the user is happy again, because they immediately see that there multiple default routes and potentially avoid major confusion. In both cases, cutting the list early sounds like premature optimization. I would be very surprised if the time we save was measurable. I would very much prefer to return a well-defined answer rather than a quick one.

(For me, the special names like `_gateway` are as often useful as a diagnostic tool, if not more often, then a mechanism to make the actual connection. Thus, complete output is more important than output that allows a connection to be made.)

keszybz commented on src/shared/local-addresses.c:
```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+
+                /* Generate exactly one entry per protocol */
+                if (done_af != AF_UNSPEC && af == done_af)
+                        continue;
```
... here

keszybz commented on src/shared/local-addresses.c:
```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+                        sa.in = (struct sockaddr_in) {
+                                .sin_family = AF_INET,
+                                .sin_addr = a[i].address.in,
+                                .sin_port = htobe16(53), /* doesn't really matter wich port we pick — we just care about the routing decision */
```
which

keszybz reviewed (commented):
Looks reasonable in general, but I'd like to pin down the semantics before merging.

I'll unset the green labels for now, as there are various outstanding comments.

poettering commented:
Force pushed a new version, mostly with a lot more comments, addressing the point @bluca and @keszybz made. ptal!

behrmann commented on man/nss-myhostname.xml:
```diff
@@ -51,6 +51,13 @@
+      <listitem><para>The hostname <literal>_outbound</literal> is resolved to the local IPv4 and IPv6
+      addresses that are most likely used for communication with other hosts. This is determined by
+      requesting a routing decision to the configured default gateway from the kernel and then using the
+      local IP address selected by this decision. This host name is only available if there is a local
```
```suggestion
      local IP address selected by this decision. This hostname is only available if there is a local
```
at least it's written like that in the test of the docs I can see here.

behrmann commented on man/systemd-resolved.service.xml:
```diff
@@ -104,6 +104,13 @@
+      <listitem><para>The hostname <literal>_outbound</literal> is resolved to the local IPv4 and IPv6
+      addresses that are most likely used for communication with other hosts. This is determined by
+      requesting a routing decision to the configured default gateway from the kernel and then using the
+      local IP address selected by this decision. This host name is only available if there is a local
```
```suggestion
      local IP address selected by this decision. This hostname is only available if there is a local
```
same here

poettering commented:
I have now force pushed a new version, addressing all issues.

I am not sure I agree with @keszybz' request to return all possible local addresses the kernel chooses for its routing decision, instead of just the ipv4 + ipv6 addresses with the highest priority. To me it defeats half the purpose of the concept, i.e. picking just the highest likely option, with no ambiguities regarding preference of the other entries makes the most sense to me, at least for my usecase, which is answering "what's *the* local IP address?". Answering that with a plural response misses the question's point, im pov.

But given this is likely a corner case, I reworked this now, in the hope we can get this merged soon.

bluca commented on src/test/test-local-addresses.c:
```diff
@@ -40,5 +40,12 @@ int main(int argc, char *argv[]) {
         free(a);
 
+        n = local_outbounds(NULL, 0, AF_UNSPEC, &a);
+        assert_se(n >= 0);
```
will this test fail if the tests are run in some unprivileged, network-less, constrained container?

poettering replied:
well, this PR doesn't add much new in that area. i.e. we first invoke local_gateways() (which the test calls anyway already), and then issue a connect() to each gateway, and ignore almost all errors on those socket operations. Thus, if the test previously was not an issue it shouldn't be an issue now either. The security-wise problematic stuff is more likely the rtnetlink stuff, but so far this proved not to be a problem. The UDP socket stuff we do should be unproblematic, in particular as we — as mentioned — ignore almost all errors for it, or handle them gracefully

bluca commented:
still looks good to me, just a question about the test - waiting for @keszybz before merging

keszybz commented:
Thanks. I think the semantics are nice:
```console
$ resolvectl query _outbound
_outbound: 192.168.1.40                        -- link: hub0
           192.168.1.33                        -- link: wlp4s0
           fe80::537:cb90:8b8c:be35%5          -- link: wlp4s0
```
I guess that there are at least two fairly common cases where this becomes relevant: a laptop with both network and wireless connectivity, and a device with VPN configured as a default route. Both cases are nicely served automatically with NetworkManager, and this code should DTRT.
